File: python/tic-tac-toe/tic_tac_toe.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
三目並べ (Tic-Tac-Toe) - Python実装
シンプルなコマンドライン版の三目並べゲーム
"""


class TicTacToe:
    def __init__(self):
        """ゲームの初期化"""
        self.board = [' ' for _ in range(9)]
        self.current_player = 'X'
        
    def display_board(self):
        """ボードを表示"""
        print("\n")
        print(f" {self.board[0]} | {self.board[1]} | {self.board[2]} ")
        print("---|---|---")
        print(f" {self.board[3]} | {self.board[4]} | {self.board[5]} ")
        print("---|---|---")
        print(f" {self.board[6]} | {self.board[7]} | {self.board[8]} ")
        print("\n")
        
    def display_positions(self):
        """位置番号を表示"""
        print("\n位置番号:")
        print(" 1 | 2 | 3 ")
        print("---|---|---")
        print(" 4 | 5 | 6 ")
        print("---|---|---")
        print(" 7 | 8 | 9 ")
        print()
        
    def is_valid_move(self, position):
        """移動が有効かチェック"""
        return 0 <= position < 9 and self.board[position] == ' '
    
    def make_move(self, position):
        """移動を実行"""
        if self.is_valid_move(position):
            self.board[position] = self.current_player
            return True
        return False
    
    def check_winner(self):
        """勝者をチェック"""
        # 勝利パターン
        winning_combinations = [
            [0, 1, 2], [3, 4, 5], [6, 7, 8],  # 横
            [0, 3, 6], [1, 4, 7], [2, 5, 8],  # 縦
            [0, 4, 8], [2, 4, 6]               # 斜め
        ]
        
        for combo in winning_combinations:
            if (self.board[combo[0]] == self.board[combo[1]] == 
                self.board[combo[2]] != ' '):
                return self.board[combo[0]]
        return None
    
    def is_board_full(self):
        """ボードが埋まっているかチェック"""
        return ' ' not in self.board
    
    def switch_player(self):
        """プレイヤーを切り替え"""
        self.current_player = 'O' if self.current_player == 'X' else 'X'
    
    def play(self):
        """ゲームを実行"""
        print("=" * 40)
        print("三目並べ (Tic-Tac-Toe) へようこそ!")
        print("=" * 40)
        self.display_positions()
        
        while True:
            self.display_board()
            print(f"プレイヤー {self.current_player} の番です")
            
            try:
                position = int(input("位置を選択 (1-9): ")) - 1
                
                if not self.is_valid_move(position):
                    print("無効な移動です。もう一度試してください。")
                    continue
                
                self.make_move(position)
                
                # 勝者チェック
                winner = self.check_winner()
                if winner:
                    self.display_board()
                    print(f"🎉 プレイヤー {winner} の勝ちです!")
                    break
                
                # 引き分けチェック
                if self.is_board_full():
                    self.display_board()
                    print("引き分けです!")
                    break
                
                # プレイヤー切り替え
                self.switch_player()
                
            except (ValueError, IndexError):
                print("1から9の数字を入力してください。")
            except KeyboardInterrupt:
                print("\n\nゲームを終了します。")
                break


def main():
    """メイン関数"""
    while True:
        game = TicTacToe()
        game.play()
        
        play_again = input("\nもう一度プレイしますか? (y/n): ").lower()
        if play_again != 'y':
            print("ありがとうございました!")
            break


if __name__ == "__main__":
    main()

File: javascript/tic-tac-toe/tic_tac_toe.js
#!/usr/bin/env node
/**
 * 三目並べ (Tic-Tac-Toe) - JavaScript実装
 * Node.jsで動作するコマンドライン版の三目並べゲーム
 */

const readline = require('readline');

class TicTacToe {
  constructor() {
    this.board = Array(9).fill(' ');
    this.currentPlayer = 'X';
    this.rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });
  }

  /**
   * ボードを表示
   */
  displayBoard() {
    console.log('\n');
    console.log(` ${this.board[0]} | ${this.board[1]} | ${this.board[2]} `);
    console.log('---|---|---');
    console.log(` ${this.board[3]} | ${this.board[4]} | ${this.board[5]} `);
    console.log('---|---|---');
    console.log(` ${this.board[6]} | ${this.board[7]} | ${this.board[8]} `);
    console.log('\n');
  }

  /**
   * 位置番号を表示
   */
  displayPositions() {
    console.log('\n位置番号:');
    console.log(' 1 | 2 | 3 ');
    console.log('---|---|---');
    console.log(' 4 | 5 | 6 ');
    console.log('---|---|---');
    console.log(' 7 | 8 | 9 ');
    console.log();
  }

  /**
   * 移動が有効かチェック
   */
  isValidMove(position) {
    return position >= 0 && position < 9 && this.board[position] === ' ';
  }

  /**
   * 移動を実行
   */
  makeMove(position) {
    if (this.isValidMove(position)) {
      this.board[position] = this.currentPlayer;
      return true;
    }
    return false;
  }

  /**
   * 勝者をチェック
   */
  checkWinner() {
    const winningCombinations = [
      [0, 1, 2], [3, 4, 5], [6, 7, 8], // 横
      [0, 3, 6], [1, 4, 7], [2, 5, 8], // 縦
      [0, 4, 8], [2, 4, 6]              // 斜め
    ];

    for (const combo of winningCombinations) {
      if (this.board[combo[0]] === this.board[combo[1]] &&
          this.board[combo[1]] === this.board[combo[2]] &&
          this.board[combo[0]] !== ' ') {
        return this.board[combo[0]];
      }
    }
    return null;
  }

  /**
   * ボードが埋まっているかチェック
   */
  isBoardFull() {
    return !this.board.includes(' ');
  }

  /**
   * プレイヤーを切り替え
   */
  switchPlayer() {
    this.currentPlayer = this.currentPlayer === 'X' ? 'O' : 'X';
  }

  /**
   * ユーザー入力を取得
   */
  async getInput(question) {
    return new Promise((resolve) => {
      this.rl.question(question, (answer) => {
        resolve(answer);
      });
    });
  }

  /**
   * ゲームメインループ
   */
  async play() {
    console.log('========================================');
    console.log('三目並べ (Tic-Tac-Toe) へようこそ!');
    console.log('========================================');
    this.displayPositions();

    while (true) {
      this.displayBoard();
      console.log(`プレイヤー ${this.currentPlayer} の番です`);

      const input = await this.getInput('位置を選択 (1-9): ');
      const position = parseInt(input) - 1;

      if (isNaN(position) || !this.isValidMove(position)) {
        console.log('無効な移動です。もう一度試してください。');
        continue;
      }

      this.makeMove(position);

      // 勝者チェック
      const winner = this.checkWinner();
      if (winner) {
        this.displayBoard();
        console.log(`🎉 プレイヤー ${winner} の勝ちです!`);
        break;
      }

      // 引き分けチェック
      if (this.isBoardFull()) {
        this.displayBoard();
        console.log('引き分けです!');
        break;
      }

      // プレイヤー切り替え
      this.switchPlayer();
    }
  }

  /**
   * リソースをクリーンアップ
   */
  close() {
    this.rl.close();
  }
}

/**
 * メイン関数
 */
async function main() {
  while (true) {
    const game = new TicTacToe();
    
    try {
      await game.play();
      
      const playAgain = await game.getInput('\nもう一度プレイしますか? (y/n): ');
      game.close();
      
      if (playAgain.toLowerCase() !== 'y') {
        console.log('ありがとうございました!');
        break;
      }
    } catch (error) {
      game.close();
      console.log('\n\nゲームを終了します。');
      break;
    }
  }
  
  process.exit(0);
}

// プログラムを実行
if (require.main === module) {
  main().catch((error) => {
    console.error('エラーが発生しました:', error);
    process.exit(1);
  });
}

module.exports = TicTacToe;
